### PR TITLE
Allow renaming of immutable secrets

### PR DESCRIPTION
**Pull request by jandubois**

### CODE REVIEW

File: secret/secret_test.go [hunk 1]
```diff
@@ -619,6 +619,89 @@ func TestUpdateVariable(t *testing.T) {
 		configVar := model.ConfigurationVariable{Name: "IN_UPDATES"}
 		result := updateVariable(&secrets, &update, &configVar)
 		assert.False(result)
-		assert.Equal(string(secrets.Data["in-updates"]), "orig")
+		assert.Equal("orig", string(secrets.Data["in-updates"]))
+	})
+}
+
+func TestMigrateRenamedVariable(t *testing.T) {
+	assert := assert.New(t)
```
Comment (mook-as) [inline secret/secret_test.go]: Is this correct? Doesn't the `assert` object need to be created from the `t` of the individual sub-tests?

File: secret/secret_test.go [hunk 2]
```diff
@@ -619,6 +619,89 @@ func TestUpdateVariable(t *testing.T) {
...
+	t.Run("NoPreviousNames", func(t *testing.T) {
+		t.Parallel()
+
+		// If `name` has no previous names, then it should remain empty and `dirty` should be false
+		secrets := v1.Secret{Data: map[string][]byte{}}
+
+		configVar := model.ConfigurationVariable{Name: "NEW_NAME"}
+		result := migrateRenamedVariable(&secrets, &configVar)
+		assert.False(result)
+		assert.Equal("", string(secrets.Data["new-name"]))
```
Comment (mook-as) [inline secret/secret_test.go]: `assert.Empty()` is an option, but I'm not sure that is actually better here…

File: secret/secret_test.go [hunk 3]
```diff
@@ -619,6 +619,89 @@ func TestUpdateVariable(t *testing.T) {
...
+	})
+
+	t.Run("PreviousNameWithoutValue", func(t *testing.T) {
+		t.Parallel()
+
+		// If `name` has a previous name, but without value, then it should remain empty and `dirty` should be false
+		secrets := v1.Secret{Data: map[string][]byte{}}
+		secrets.Data["previous-name"] = []byte("")
```
Comment (mook-as) [inline secret/secret_test.go]: Writing things out to see how nice they look…
```go
secrets := v1.Secret{Data: map[string][]byte{"previous-name": []byte("")}}
```
Not really better, I guess.

File: ssh/ssh_test.go [hunk 4]
```diff
@@ -34,7 +34,7 @@ func TestNewKeyIsCreated(t *testing.T) {
 
 	// 16 colon separated bytes = 47
 	// 00:11:22:33:44:55:66:77:88:99:aa:bb:cc:dd:ee:ff
-	assert.Equal(len(secretData["bar"]), 47)
+	assert.Equal(47, len(secretData["bar"]))
```
Comment (mook-as) [inline ssh/ssh_test.go]: `assert.Len()` again.

File: ssl/ssl_test.go [hunk 5]
```diff
@@ -383,16 +383,16 @@ func TestAddHost(t *testing.T) {
 	t.Run("Only host is added when wildcard is false", func(t *testing.T) {
 		req := csr.CertificateRequest{}
 		addHost(&req, false, "name")
-		assert.Equal(req.Hosts[0], "name")
-		assert.Equal(len(req.Hosts), 1)
+		assert.Equal("name", req.Hosts[0])
+		assert.Equal(1, len(req.Hosts))
```
Comment (mook-as) [inline ssl/ssl_test.go]: `assert.Len()`… also once a few lines down.

File: secret/secret.go [hunk 6]
```diff
@@ -133,19 +135,33 @@ func GenerateSecrets(manifest model.Manifest, secrets *v1.Secret, updates *v1.Se
...
 
-func updateVariable(secrets *v1.Secret, updates *v1.Secret, configVar *model.ConfigurationVariable) (dirty bool) {
+func updateVariable(secrets, updates *v1.Secret, configVar *model.ConfigurationVariable) {
 	name := util.ConvertNameToKey(configVar.Name)
 	if len(secrets.Data[name]) == 0 && len(updates.Data[name]) > 0 {
 		secrets.Data[name] = updates.Data[name]
-		dirty = true
+		util.MarkAsDirty(secrets)
 	}
 	return
```
Comment (mook-as) [inline secret/secret.go]: Huh, a lone `return` by itself at the end of a void function doesn't look very useful.

File: secret/secret_test.go [hunk 7]
```diff
@@ -120,58 +119,52 @@ func (m *MockSecrets) RecordSSLCertInfo(configVar *model.ConfigurationVariable)
...
-	return results.Bool(0)
+func (m *MockSecrets) GenerateSSLCerts(secrets, updates *v1.Secret) {
+	m.Called(secrets, updates)
 }
 
 func TestUpdateSecretsWhenCreatingOrUpdating(t *testing.T) {
 	t.Parallel()
 
 	t.Run("Neither Create Nor Update called", func(t *testing.T) {
 		var s MockSecretInterface
```
Comment (mook-as) [inline secret/secret_test.go]: Should there be a `t.Parallel()` call here? (So that the various subtests can be run parallel to each other)

File: secret/secret_test.go [hunk 8]
```diff
@@ -120,58 +119,52 @@ func (m *MockSecrets) RecordSSLCertInfo(configVar *model.ConfigurationVariable)
...
 
-		UpdateSecrets(&s, nil, false, false)
-		s.AssertNotCalled(t, "Create", nil)
-		s.AssertNotCalled(t, "Update", nil)
+		s.On("Create", secrets).Return(nil, nil)
+		s.On("Update", secrets).Return(nil, nil)
+
+		UpdateSecrets(&s, secrets, false)
+		s.AssertNotCalled(t, "Create", secrets)
+		s.AssertNotCalled(t, "Update", secrets)
```
Comment (mook-as) [inline secret/secret_test.go]: Assert that the secrets are not dirty? That it's empty?

File: secret/secret_test.go [hunk 9]
```diff
@@ -120,58 +119,52 @@ func (m *MockSecrets) RecordSSLCertInfo(configVar *model.ConfigurationVariable)
...
-		s.On("Update", (*v1.Secret)(nil)).Return(nil, nil)
+		s.On("Create", secrets).Return(nil, nil)
+		s.On("Update", secrets).Return(nil, nil)
 
-		UpdateSecrets(&s, nil, true, true)
-		s.AssertCalled(t, "Create", (*v1.Secret)(nil))
-		s.AssertNotCalled(t, "Update", (*v1.Secret)(nil))
+		UpdateSecrets(&s, secrets, true)
+		s.AssertCalled(t, "Create", secrets)
+		s.AssertNotCalled(t, "Update", secrets)
```
Comment (mook-as) [inline secret/secret_test.go]: Assert that it only has the expected secrets in it? (I.e. that it's not dirty)

Same for the next one.

File: secret/secret_test.go [hunk 10]
```diff
@@ -224,8 +217,8 @@ func TestGetOrCreateWithValidSecrets(t *testing.T) {
 		s.On("Get", SECRET_NAME, metav1.GetOptions{})
 		create, secrets, _ := GetOrCreateSecrets(&s)
 		s.AssertCalled(t, "Get", SECRET_NAME, metav1.GetOptions{})
-		assert.Equal(secrets.Data[SECRET_NAME], []byte(SECRET_NAME))
-		assert.False(create)
+		assert.Equal(t, []byte(SECRET_NAME), secrets.Data[SECRET_NAME])
+		assert.False(t, create)
```
Comment (mook-as) [inline secret/secret_test.go]: More descriptive messages would be nice.

File: secret/secret_test.go [hunk 11]
```diff
@@ -277,19 +268,6 @@ func TestGenerateSecretsWithNoSecrets(t *testing.T) {
 		generateSSLCerts = origGenerateSSLCerts
 	}()
 
-	t.Run("Manifest with no secrets doesn't change", func(t *testing.T) {
```
Comment (mook-as) [inline secret/secret_test.go]: We no longer need to assert this?

File: secret/secret_test.go [hunk 12]
```diff
@@ -305,17 +283,16 @@ func TestGenerateSecretsWithNoSecrets(t *testing.T) {
...
 		passGenerate = mockSecrets.PassGenerate
-		mockSecrets.On("GenerateSSLCerts", &secrets, &updates).Return(false)
-		mockSecrets.On("PassGenerate", map[string][]byte{}, map[string][]byte{}, "dirty").Return(true)
-		dirty := GenerateSecrets(manifest, &secrets, &updates)
-		assert.True(dirty)
-		mockSecrets.AssertCalled(t, "PassGenerate", map[string][]byte{}, map[string][]byte{}, "dirty")
+		mockSecrets.On("GenerateSSLCerts", secrets, updates)
+		mockSecrets.On("PassGenerate", secrets, updates, "dirty")
+		GenerateSecrets(manifest, secrets, updates)
+		mockSecrets.AssertCalled(t, "PassGenerate", secrets, updates, "dirty")
```
Comment (mook-as) [inline secret/secret_test.go]: Assert that the secrets are dirty?

File: secret/secret_test.go [hunk 13]
```diff
@@ -333,17 +310,17 @@ func TestGenerateSecretsWithNoSecrets(t *testing.T) {
...
-		mockSecrets.On("GenerateSSLCerts", &secrets, &updates).Return(false)
-		mockSecrets.On("PassGenerate", map[string][]byte{}, map[string][]byte{}, "clean").Return(false)
-		dirty := GenerateSecrets(manifest, &secrets, &updates)
-		assert.False(dirty)
-		mockSecrets.AssertCalled(t, "PassGenerate", map[string][]byte{}, map[string][]byte{}, "clean")
+		mockSecrets.On("GenerateSSLCerts", secrets, updates)
+		mockSecrets.On("PassGenerate", secrets, updates, "clean")
+
+		GenerateSecrets(manifest, secrets, updates)
+		mockSecrets.AssertCalled(t, "PassGenerate", secrets, updates, "clean")
```
Comment (mook-as) [inline secret/secret_test.go]: Assert that it's _not_ dirty? (Also the next one; similar for more removal of asserting true/false.)

File: secret/secret_test.go [hunk 14]
```diff
@@ -6,6 +6,7 @@ import (
```
Comment (mook-as) [inline secret/secret_test.go]: We should probably test that `UpdateSecrets()` clears the dirty flag (and that it's never actually written to the secret store).

File: ssl/ssl_test.go [hunk 15]
```diff
@@ -158,22 +153,20 @@ func TestGenerateCerts(t *testing.T) {
...
-		updates := v1.Secret{Data: map[string][]byte{}}
+		secrets := &v1.Secret{Data: map[string][]byte{}}
+		updates := &v1.Secret{Data: map[string][]byte{}}
 
 		// When createCA returns true
-		mockSSL.On("createCA", &secrets, &updates, CERT_ID).Return(true)
-		dirty := GenerateCerts(&secrets, &updates)
-		assert.True(dirty)
-		mockSSL.AssertCalled(t, "createCA", &secrets, &updates, CERT_ID)
+		mockSSL.On("createCA", secrets, updates, CERT_ID).Return(true)
```
Comment (mook-as) [inline ssl/ssl_test.go]: Does this function still return anything? Also the same call a few lines later.

File: ssl/ssl_test.go [hunk 16]
```diff
@@ -187,22 +180,20 @@ func TestGenerateCerts(t *testing.T) {
...
-		updates := v1.Secret{Data: map[string][]byte{}}
+		secrets := &v1.Secret{Data: map[string][]byte{}}
+		updates := &v1.Secret{Data: map[string][]byte{}}
 
 		// When createCert returns true
-		mockSSL.On("createCert", &secrets, &updates, CERT_ID).Return(true)
-		dirty := GenerateCerts(&secrets, &updates)
-		assert.True(dirty)
-		mockSSL.AssertCalled(t, "createCert", &secrets, &updates, CERT_ID)
+		mockSSL.On("createCert", secrets, updates, CERT_ID).Return(true)
```
Comment (mook-as) [inline ssl/ssl_test.go]: This shouldn't return anything either (see also a few lines later)

File: password/password_test.go [hunk 17]
```diff
@@ -3,32 +3,30 @@ package password
...
 
-	secretData := make(map[string][]byte)
-	updateData := make(map[string][]byte)
+	GeneratePassword(secrets, updates, "foo")
 
-	result := GeneratePassword(secretData, updateData, "foo")
-
-	assert.True(result)
-	assert.Equal(len(secretData["foo"]), 64)
+	assert.True(t, util.IsDirty(secrets))
```
[the remark this answers is not in the document]
Comment (jandubois) [reply password/password_test.go]: Ok, will add them here, but not everywhere else. :)

I find the additional text not that helpful, as you already get the name of the failing test function as well as the filename and line number...

File: secret/secret_test.go (hunk 8 above)
Comment (jandubois) [reply secret/secret_test.go]: I don't know; `UpdateSecrets` doesn't do anything to the `secrets` in-memory structure except for resetting the dirty flag.

There are endless things that should *not* happen during the call; why do you want to test for this particular thing? Same for the other calls to `UpdateSecrets` (except that I did mark the secrets as `dirty` and now check that they are clean when the function returns).

File: secret/secret_test.go (hunk 11 above)
Comment (jandubois) [reply secret/secret_test.go]: That test is passing in `nil` as the `secrets` and `updates`, which aren't really valid values to begin with.

We could add a real test that an empty manifest doesn't change anything, but feels like low-value to me.

File: secret/secret_test.go (hunk 8 above)
Comment (mook-as) [reply secret/secret_test.go]: Mostly, I just wanted something to ensure that we don't leak the dirtiness flag into the actual secret store. I'm happy however you desire to check that.

File: secret/secret_test.go (hunk 12 above)
Comment (jandubois) [reply secret/secret_test.go]: They won't be because we have mocked `PassGenerate`, which would be setting the dirty flag. This is a test for `GenerateSecrets` which doesn't touch the `dirty` flag, so there is no reason to test for it here. Same for the next comment.

**Comment by mook-as** [comment]
Sorry for the delay, finally managed to *correctly* deploy this on a vagrant box and observe the secret changing (at least, I think I did).
